std/zig/parse.zig

```diff
@@ -1,3659 +1,3044 @@
-                    Token.Id.Bang => {
-                        fn_proto.return_type = ast.Node.FnProto.ReturnType{ .InferErrorSet = undefined };
-                        stack.append(State{ .TypeExprBegin = OptionalCtx{ .Required = &fn_proto.return_type.InferErrorSet } }) catch unreachable;
-                        continue;
-                    },
-                    else => {
-                        // TODO: this is a special case. Remove this when #760 is fixed
-                        if (token_ptr.id == Token.Id.Keyword_anyerror) {
-                            if (tok_it.peek().?.id == Token.Id.LBrace) {
-                                const error_type_node = try arena.create(ast.Node.ErrorType);
-                                error_type_node.* = ast.Node.ErrorType{
-                                    .base = ast.Node{ .id = ast.Node.Id.ErrorType },
-                                    .token = token_index,
-                                };
-                                fn_proto.return_type = ast.Node.FnProto.ReturnType{ .Explicit = &error_type_node.base };
-                                continue;
-                            }
-                        }
-
-                        prevToken(&tok_it, tree);
-                        fn_proto.return_type = ast.Node.FnProto.ReturnType{ .Explicit = undefined };
-                        stack.append(State{ .TypeExprBegin = OptionalCtx{ .Required = &fn_proto.return_type.Explicit } }) catch unreachable;
-                        continue;
-                    },
-                }
-            },
+// VarDecl <- (KEYWORD_const / KEYWORD_var) IDENTIFIER (COLON TypeExpr)? ByteAlign? LinkSection? (EQUAL Expr)? SEMICOLON
+fn parseVarDecl(arena: *Allocator, it: *TokenIterator, tree: *Tree) !?*Node {
+    const mut_token = eatToken(it, .Keyword_const) orelse
+        eatToken(it, .Keyword_var) orelse
+        return null;
+
+    const name_token = try expectToken(it, tree, .Identifier);
+    const type_node = if (eatToken(it, .Colon) != null)
+        try expectNode(arena, it, tree, parseTypeExpr, AstError{
+            .ExpectedTypeExpr = AstError.ExpectedTypeExpr{ .token = it.index },
+        })
+    else
+        null;
+    const align_node = try parseByteAlign(arena, it, tree);
+    const section_node = try parseLinkSection(arena, it, tree);
+    const eq_token = eatToken(it, .Equal);
+    const init_node = if (eq_token != null) blk: {
+        break :blk try expectNode(arena, it, tree, parseExpr, AstError{
+            .ExpectedExpr = AstError.ExpectedExpr{ .token = it.index },
+        });
+    } else null;
+    const semicolon_token = try expectToken(it, tree, .Semicolon);
+
+    const node = try arena.create(Node.VarDecl);
+    node.* = Node.VarDecl{
+        .base = Node{ .id = .VarDecl },
+        .doc_comments = null,
+        .visib_token = null,
+        .thread_local_token = null,
+        .name_token = name_token,
+        .eq_token = eq_token orelse 0,
```

Comment:
Oops I missed this. Yes, this should probably be, above, `const eq_token = eatToken(it, .Equal) orelse (return a parse error)`

Reply:
Good catch, I completely forgot to leave a `TODO` -- and still getting used to treating `undefined` as a debug tool.

I see what happened here. I didn't use `expectToken` because the grammar shows the assignment is optional: `(EQUAL Expr)?`. Presumably this is to support externs, which allow variable decls with no initialization (`extern const food: SomeType;`), unlike regular variables. I'd grepped the repo for such statements and found nothing, but now I get it.

Between making `eq_token` optional and leaving it undefined, in this case I think I'd prefer the latter. Less impact on other code, and it's only accessed when `init_node` is present: https://github.com/ziglang/zig/blob/master/std/zig/render.zig#L1801-L1804

Reply:
Ahh, I apologize, I misunderstood and gave bad advice. I see now the issue, and I think your proposed solution is fine. The other, more destructive but perhaps better data modeling approach, would be:

```diff
--- a/std/zig/ast.zig
+++ b/std/zig/ast.zig
@@ -508,7 +508,6 @@ pub const Node = struct {
         visib_token: ?TokenIndex,
         thread_local_token: ?TokenIndex,
         name_token: TokenIndex,
-        eq_token: TokenIndex,
         mut_token: TokenIndex,
         comptime_token: ?TokenIndex,
         extern_export_token: ?TokenIndex,
@@ -516,9 +515,14 @@ pub const Node = struct {
         type_node: ?*Node,
         align_node: ?*Node,
         section_node: ?*Node,
-        init_node: ?*Node,
+        init: ?Init,
         semicolon_token: TokenIndex,
 
+        pub const Init = struct {
+            eq_token: TokenIndex,
+            expr: *Node,
+        };
+
```

I think your approach sounds good, we can do another pass on the ast structures later and see if we want to do these kinds of data modeling improvements.